### PR TITLE
[UI] Show winetricks progress output and Sync-Saves output via React Dialog

### PR DESCRIPTION
**Pull request by Nocccer**
- Show winetricks output on frontend to let the user know if winetricks is still running or stopped working.
- If winetricks throws an error we will show the error in frontend aswell.

https://user-images.githubusercontent.com/61798668/184673791-7999db4a-5057-4f07-8d02-4666afdfcd78.mp4

- Sync-Saves also using the same React component instead electron dialog

https://user-images.githubusercontent.com/61798668/186752764-7b3e3a87-a2ca-4351-8ebf-0abce14366ed.mp4

---

Use the following Checklist if you have changed something on the Backend or Frontend:

- [x] Tested the feature and it's working on a current and clean install.
- [ ] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

### CODE REVIEW

File: src/screens/Settings/components/Tools/index.tsx [hunk 1]
```diff
@@ -18,6 +20,14 @@ export default function Tools({ appName, runner }: Props) {
...
   const [winecfgRunning, setWinecfgRunning] = useState(false)
   const [winetricksRunning, setWinetricksRunning] = useState(false)
+  const [progress, setProgress] = useState<string[]>([])
+  const winetricksOutputBottomRef = useRef<HTMLDivElement>(null)
+  const logRef = useRef<HTMLDivElement>(null)
+  const [autoScroll, setAutoScroll] = useState(true)
+
+  ipcRenderer.on('progressOfWinetricks', (e, messages) => {
+    setProgress(messages)
+  })
```
Comment (arielj) [inline src/screens/Settings/components/Tools/index.tsx:30]: I think we have to do this in a `useEffect(() => ...., [])` so we can remove the listener when this is unmounted

File: electron/tools.ts [hunk 2]
```diff
@@ -205,39 +205,88 @@ export const Winetricks = {
...
 
-    child.on('error', (error) => {
-      logError(`Winetricks throwed Error: ${error}`, LogPrefix.WineTricks)
+      child.on('error', (error) => {
+        logError(`Winetricks throwed Error: ${error}`, LogPrefix.WineTricks)
+        showErrorBoxModalAuto(
+          i18next.t('box.error.winetricks.title', 'Winetricks error'),
+          i18next.t('box.error.winetricks.message', {
+            defaultValue:
+              'Winetricks throwed following error during execution:{{newLine}}{{error}}',
```
Comment (arielj) [inline electron/tools.ts:272]:
```suggestion
              'Winetricks returned the following error during execution:{{newLine}}{{error}}',
```

I think the idea of `throwing` errors is clear for developers but not that friendly for users

File: electron/tools.ts [hunk 3]
```diff
@@ -205,39 +205,88 @@ export const Winetricks = {
...
+      child.stderr.setEncoding('utf8')
+      child.stderr.on('data', (data: string) => {
+        logError(data, LogPrefix.WineTricks)
+        appendMessage(data)
+      })
 
-    child.on('error', (error) => {
-      logError(`Winetricks throwed Error: ${error}`, LogPrefix.WineTricks)
+      child.on('error', (error) => {
+        logError(`Winetricks throwed Error: ${error}`, LogPrefix.WineTricks)
```
Comment (arielj) [inline electron/tools.ts:267]:
```suggestion
        logError(`Winetricks threw Error: ${error}`, LogPrefix.WineTricks)
```

File: src/screens/Settings/components/Tools/index.tsx [hunk 4]
```diff
@@ -79,6 +119,50 @@ export default function Tools({ appName, runner }: Props) {
...
+          >
+            <DialogHeader
+              onClose={() => {
+                return
+              }}
+            >
+              <div>Winetricks</div>
+            </DialogHeader>
+            <DialogContent>
+              <div>Progress:</div>
```
Comment (arielj) [inline src/screens/Settings/components/Tools/index.tsx:137]:
```suggestion
              <div>{t('progress', 'Progress:')}</div>
```

**Review by arielj — COMMENTED** [review]
looks good and works good too! I left some comments

File: src/screens/Settings/components/Tools/index.tsx [hunk 5]
```diff
@@ -38,6 +44,42 @@ export default function Tools({ appName, runner }: Props) {
     setWinecfgRunning(false)
   }
 
+  useEffect(() => {
+    ipcRenderer.on('progressOfWinetricks', (e, messages) => {
+      setProgress(messages)
+    })
+  }, [])
```
Comment (arielj) [inline src/screens/Settings/components/Tools/index.tsx:51]: To remove the listener on unmount you have to return a `cleanup` function:

```ts
  useEffect(() => {
    const onProgress = (e, messages) => {
      setProgress(messages)
    }

    ipcRenderer.on('progressOfWinetricks', onProgress)

    () => ipcRenderer.removeListener('progressOfWinetricks', onProgress)
  }, [])
```

**Review by arielj — APPROVED** [review]
looks good to me, just left a comment on the cleanup function if the ipcRenderer.on

I tested installing some thing and things worked as expected

**Review by flavioislima — APPROVED** [review]
Pretty good stuff.
Tested here with both Wine and Proton prefixes and works like charm ⚔️

**Pull request merged**